### ISSUE TITLE
Authenticated HTTPS proxy config is not applied in `http_archive` downloads

### ISSUE THREAD

**Issue by timsutton** [bazelbuild/bazel#14675]
### Description of the problem / feature request:

I'm having trouble with Bazel retrieving a resource defined by an `http_archive` repository rule, through an authenticated proxy. I'm on a system in which we would need to retrieve any public resources through a proxy that also requires authentication, and am trying to build envoy as a test case.

My configuration for actually routing through the proxy seems to be getting picked up, but I receive a 407 (proxy authentication required) error and can confirm on the proxy side that it receives no user/password auth.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I don't have at the moment a self-contained example of a local proxy configured to require authentication, to attach here. However I suspect I can glean more about what's going on if I knew what additional debug/log flags I could potentially pass in.

My test case:

```bash
git clone https://github.com/envoyproxy/envoy && cd envoy

# dummy values substituted for my actual user/pass/host/port
bazelisk \
'--host_jvm_args=-Dhttps.proxyHost=https://myuser:mypass@my.proxy.host' \
'--host_jvm_args=-Dhttps.proxyPort=1234' \
build //source/exe:envoy-static
```

Eventually I'll get this error:

```
ERROR: no such package '@bazel_gazelle//': java.io.IOException: Error downloading [https://github.com/bazelbuild/bazel-gazelle/releases/download/v0.24.0/bazel-gazelle-v0.24.0.tar.gz] to /private/var/tmp/_bazel_ec2-user/3fdb6675002b9d71cdd6dd7b0c48f6e7/external/bazel_gazelle/temp12986163671727929748/bazel-gazelle-v0.24.0.tar.gz: Unable to tunnel through proxy. Proxy returns "HTTP/1.1 407 Proxy Authentication Required"
```

I've tried what should be an equivalent, of just setting HTTPS_PROXY:

```
export HTTPS_PROXY='https://myuser:mypass@my.proxy.host:1234'
bazelisk build //source/exe:envoy-static
```

### What operating system are you running Bazel on?

macOS Monterey, 12.1.

### What's the output of `bazel info release`?

```
release 4.2.1 
```

###  Have you found anything relevant by searching the web?

I've found a few issues referencing authenticated proxies but nothing particularly helpful. I saw that the support for honoring the system properties for proxies was added in https://github.com/bazelbuild/bazel/pull/11944, and originally adding proxy auth support was added back in https://github.com/bazelbuild/bazel/issues/587. 

### Any other information, logs, or outputs that you want to share?

To try and ensure my user/pass was getting properly parsed out of the https:// URL regex [here](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/ProxyHelper.java#L136), I tried a few iterations of my `https.proxyHost`, of leaving out the password, or mangling the proxy port, just to ensure that I was in fact walking [these code paths](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/ProxyHelper.java#L170-L180), and I was able to trigger those error conditions.

My familiarity with Java in general and of this codebase is low: but mainly it seemed like I'd need to understand whether the `Authenticator.setDefault()` and password auth that [is being configured](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/ProxyHelper.java#L184-L192) is in fact being still used to apply to the proxy instance.

I also had originally tried putting in my user auth info into a ~/.netrc file, but this didn't seem to help. Not sure that would have helped for the proxy connection anyway, or just the url specified in the http_archive instance.

I also took a look at the user output root's `java.log` from this build, and see a longer traceback leading up to the 407 error. Posted that log [here](https://gist.github.com/timsutton/b09aebc02a9e1437da8f92b0c5f0b6c5).

As someone not experienced with Java debugging: Is there additional logging I could set (for example in `java.net`) that might help me see more of what's going on, and also to not collapse the additional lines from the stacktrace?

I did also verify that I the proxy authentication for that resource using `curl` at least gets me to where I receive a 302 redirect rather than the 407 error:

```
# curl wants no https:// in https_proxy env var (I think this is more common than requiring it, but Bazel would want it currently)
$ https_proxy=myuser:mypass@my.proxy.host:1234 \
  curl -v https://github.com/bazelbuild/rules_go/releases/download/v0.29.0/rules_go-v0.29.0.zip

[snip]
< HTTP/2 302 
[snip]
```

**Comment by timsutton** [comment]
Linking in https://github.com/bazelbuild/bazel/issues/7487#issuecomment-642441599 to this issue, which looks promising.

**Comment by timsutton** [comment]
I can confirm that did it. I needed to take extra care to unset `http_proxy` and `https_proxy` environment variables that I was also using for other components of our build. The Java proxy code seems to handle parsing those find as an alternative to setting the system properties, however for those it seems to expect the URL scheme as part of the value while other tools such as `curl` explicitly don't want it, so I'm relying on only the system property instead since I can more logically separate the two.

My final example looks like:

```
bazelisk \
'--host_jvm_args=-Dhttps.proxyHost=https://myuser:mypass@my.proxy.host' \
'--host_jvm_args=-Dhttps.proxyPort=1234' \
'--host_jvm_args=-Djdk.http.auth.tunneling.disabledSchemes=' \
build //source/exe:envoy-static
```

**Comment by puneet336** [comment]
I’ve been trying to compile Bazel (as a TensorFlow dependency) and spent the last 3 hours stuck on proxy issues.

Could someone clearly explain what needs to be changed to make Bazel work behind a proxy?

Also, why do we still rely on external URLs for downloading dependencies? Why not provide a larger archive that already includes all required dependencies, so we don’t have to download them separately?

I don’t know Bazel and don’t intend to learn it in depth, but TensorFlow forces me to interact with it.

Here’s what I first tried in `~/.bazelrc`:

```text
build --host_jvm_args=-Dhttps.proxyHost=http://proxy.eu.net
build --host_jvm_args=-Dhttps.proxyPort=2010
build --host_jvm_args=-Djdk.http.auth.tunneling.disabledSchemes=
build --insecure
```

This resulted in:

```text
WARNING: Output base '/home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/tmpx98_i0uc-bazel-root/a3a35b2635ae2fc62b3fccc04f214f84' is on NFS. This may lead to surprising failures and undetermined behavior.
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
INFO: Reading rc options for 'test' from /home/singhpuv/.bazelrc:
  Inherited 'build' options: --host_jvm_args=-Dhttps.proxyHost=http://proxy.eu.novartis.net --host_jvm_args=-Dhttps.proxyPort=2010 --host_jvm_args=-Djdk.http.auth.tunneling.disabledSchemes= --insecure
ERROR: --host_jvm_args=-Dhttps.proxyHost=http://proxy.eu.novartis.net :: Unrecognized option: --host_jvm_args=-Dhttps.proxyHost=http://proxy.eu.novartis.net
```

It turned out the workaround was to use `startup` options instead of `build` options:

```text
startup --host_jvm_args=-Dhttps.proxyHost=http://proxy.eu.net
startup --host_jvm_args=-Dhttps.proxyPort=2010
startup --host_jvm_args=-Djdk.http.auth.tunneling.disabledSchemes=
startup --insecure
```

After that, files started appearing in:

```text
/home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/src/derived/distdir/
/home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/src/derived/distdir/bazel-toolchains-3.1.0.tar.gz
```

But the build still failed with:

```text
INFO: Reading 'startup' options from /home/singhpuv/.bazelrc: --host_jvm_args=-Dhttps.proxyHost=http://proxy.eu.novartis.net --host_jvm_args=-Dhttps.proxyPort=2010 --host_jvm_args=-Djdk.http.auth.tunneling.disabledSchemes= --insecure
[INFO 13:22:30.070 src/main/cpp/option_processor.cc:395] Looking for the following rc files: /etc/bazel.bazelrc,/home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/src/.bazelrc,/home/singhpuv/.bazelrc
[INFO 13:22:30.070 src/main/cpp/rc_file.cc:56] Parsing the RcFile /home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/src/.bazelrc
[INFO 13:22:30.071 src/main/cpp/rc_file.cc:56] Parsing the RcFile user.bazelrc
[INFO 13:22:30.071 src/main/cpp/rc_file.cc:129] Skipped optional import of user.bazelrc, the specified rc file either does not exist or is not readable.
[INFO 13:22:30.071 src/main/cpp/rc_file.cc:56] Parsing the RcFile /home/singhpuv/.bazelrc
[FATAL 13:22:30.075 src/main/cpp/blaze.cc:1293] Unknown startup option: '--insecure'.
  For more info, run 'bazel help startup_options'.
```

So the final working settings  `~/.bazelrc`  i thought were:

```text
startup --host_jvm_args=-Dhttps.proxyHost=http://proxy.eu.net
startup --host_jvm_args=-Dhttps.proxyPort=2010
startup --host_jvm_args=-Djdk.http.auth.tunneling.disabledSchemes=
```


but run still fails with :
```
ERROR: Shell command failed!
    full command              ->  output/bazel --output_user_root=/home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/tmpawoy01x0-bazel-root test --jobs=16 --host_javabase=@local_jdk//:jdk --subcommands --verbose_failures --build_tests_only -- //examples/cpp:hello-success_test //examples/py/... //examples/py_native:test //examples/shell/...
    exit code                 ->  1
    called from               ->  'test_step' function in /CHBS/apps/EB/software/EasyBuild/5.1.0/lib/python3.9/site-packages/easybuild/easyblocks/b/bazel.py (line 217)
    working directory         ->  /home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/src

```


```
epository rule http_archive defined at:
  /home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/src/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
WARNING: Download from https://mirror.bazel.build/github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz failed: class javax.net.ssl.SSLHandshakeException PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
WARNING: Download from https://github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz failed: class javax.net.ssl.SSLHandshakeException PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
ERROR: An error occurred during the fetch of repository 'bazel_toolchains':
   Traceback (most recent call last):
        File "/home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/src/tools/build_defs/repo/http.bzl", line 111, column 45, in _http_archive_impl
                download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://mirror.bazel.build/github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz, https://github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz] to /home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/tmpawoy01x0-bazel-root/a3a35b2635ae2fc62b3fccc04f214f84/external/bazel_toolchains/temp6832115700116342009/bazel-toolchains-3.1.0.tar.gz: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
ERROR: no such package '@bazel_toolchains//rules': java.io.IOException: Error downloading [https://mirror.bazel.build/github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz, https://github.com/bazelbuild/bazel-toolchains/releases/download/3.1.0/bazel-toolchains-3.1.0.tar.gz] to /home/singhpuv/MySoftwares/TF/buildpath/Bazel/3.7.2/GCCcore-10.2.0/tmpawoy01x0-bazel-root/a3a35b2635ae2fc62b3fccc04f214f84/external/bazel_toolchains/temp6832115700116342009/bazel-toolchains-3.1.0.tar.gz: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
INFO: Elapsed time: 31.352s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
ERROR: Couldn't start the build. Unable to run tests
FAILED: Build did NOT complete successfully (0 packages loaded)
```




I have `HTTP_PROXY`, `HTTPS_PROXY`, `http_proxy`, and `https_proxy` set in my environment, but Bazel still seems to ignore these proxy settings.

I can download files with `wget` without any issues.

At this point I’d even be willing to manually copy thousands of packages, no matter how long it takes, but I need guidance: which parts of Bazel’s code (or which configuration files) would I have to modify so that it respects the proxy or allows me to use local copies of all required dependencies?

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 8.6.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/8.6.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=8.6.0rc1. Thanks!

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 9.1.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/9.1.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=9.1.0rc1`. Thanks!
